renamer_main.cc

```diff
@@ -0,0 +1,316 @@
+                        dyn->d_un.d_val = strtab_builder.Add(bin->Str(dyn->d_un.d_val));
+                        break;
+                    default:
+                        break;
+                }
+            }
+        }
+    }
+
+    // Rewrite strings in version information
+    for (int index : sym_indecies) {
+        if (bin->verneed() && bin->versym() && !is_special_ver_ndx(bin->versym()[index])) {
+            Elf_Verneed* vn = bin->verneed();
+            for (int i = 0; i < bin->verneednum(); ++i) {
+                LOG(INFO) << "Elf_Verneed: " << SOLD_LOG_KEY(vn->vn_version) << SOLD_LOG_KEY(vn->vn_cnt)
+                          << SOLD_LOG_KEY(bin->Str(vn->vn_file)) << SOLD_LOG_KEY(vn->vn_aux) << SOLD_LOG_KEY(vn->vn_next);
+                vn->vn_file = strtab_builder.Add(bin->Str(vn->vn_file));
+                Elf_Vernaux* vna = (Elf_Vernaux*)((char*)vn + vn->vn_aux);
+                for (int j = 0; j < vn->vn_cnt; ++j) {
+                    LOG(INFO) << "Elf_Vernaux: " << SOLD_LOG_KEY(vna->vna_hash) << SOLD_LOG_KEY(vna->vna_flags)
+                              << SOLD_LOG_KEY(vna->vna_other) << SOLD_LOG_KEY(bin->strtab() + vna->vna_name) << SOLD_LOG_KEY(vna->vna_next);
+                    vna->vna_name = strtab_builder.Add(bin->Str(vna->vna_name));
+
+                    vna = (Elf_Vernaux*)((char*)vna + vna->vna_next);
+                }
+                vn = (Elf_Verneed*)((char*)vn + vn->vn_next);
+            }
+        }
+    }
+
+    // Calculate vaddrs
+    strtab_builder.Freeze();
+    Elf_Addr strtab_vaddr = 0;
+    for (const Elf_Phdr* p : bin->phdrs()) {
+        strtab_vaddr = std::max(strtab_vaddr, AlignNext(p->p_vaddr + p->p_memsz));
+    }
+    Elf_Addr gnu_hash_vaddr = strtab_vaddr + strtab_builder.size();
+
+    Elf_GnuHash gnu_hash;
+    gnu_hash.nbuckets = 1;
+    gnu_hash.symndx = 1;
+    gnu_hash.maskwords = 1;
+    gnu_hash.shift2 = 1;
+
+    Elf_Addr strtab_fileoffset = AlignNext(bin->filesize());
+    Elf_Addr gnu_hash_fileoffset = strtab_fileoffset + strtab_builder.size();
+    Elf_Addr gnu_hash_filesize = (sizeof(uint32_t) * 4 + sizeof(Elf_Addr) + sizeof(uint32_t) * (1 + sym_names.size() - gnu_hash.symndx));
+    Elf_Addr remaining_pad_fileoffset = gnu_hash_fileoffset + gnu_hash_filesize;
+    Elf_Addr strs_size = AlignNext(strtab_builder.size() + gnu_hash_filesize);  // Sum of dynstrtab and hashes
+    Elf_Addr eof_fileoffset = strtab_fileoffset + strs_size;
+
+    // Construct Phdr for new strs
+    Elf_Phdr str_phdr;
+    str_phdr.p_offset = strtab_fileoffset;
+    str_phdr.p_flags = PF_R;
+    str_phdr.p_vaddr = strtab_vaddr;
+    str_phdr.p_paddr = strtab_vaddr;
+    str_phdr.p_memsz = strs_size;
+    str_phdr.p_filesz = strs_size;
+    str_phdr.p_type = PT_LOAD;
+    str_phdr.p_align = 0x1000;
+
+    // Rewrite addresses
+    if (phdr_insert_strategy == PhdrInsertStrategy::kPT_NOTE) {
+        Elf_Phdr* victim_note = nullptr;
+        for (const Elf_Phdr* p : bin->phdrs()) {
```

Comment:
Remove `const` and you can stop using `const_cast`?

Reply:
I removed const cast by using `bin->phdrs_mut`.

Reply:
I think you didn't need `phdrs_mut` here because `phdrs` returns `const std::vector<Elf_Phdr*>`, not `const std::vector<const Elf_Phdr*>`. But yeah, using `phdrs_mut` would be more preferrable as it's now explicit we are modifying the internal structure.

Reply:
Oops, I fixed unthinkingly...